File: thread-pool.go
package main

import (
	"fmt"
	"sync"
)

type Task func()

type ThreadPool struct {
	corePoolSize int			// 核心线程数
	maxPoolSize  int			// 最大线程数
	taskQueue    chan Task		// 任务队列
	workerQueue	 chan struct{}	// 线程队列
	mu           sync.Mutex		// 锁，用于并发控制
	wg			 sync.WaitGroup // 类似信号量，避免spin带来的资源损耗
	shutdown     chan struct{}	// 关闭信号
}

// NewThreadPool 创建线程池
func NewThreadPool(corePoolSize int, maxPoolSize int, queueSize int) *ThreadPool {
	// 1. 输入参数校验
	if corePoolSize < 1 || maxPoolSize < 1 || queueSize < 0 {
		panic("输入线程数和队列大小不合法")
	}
	if corePoolSize > maxPoolSize {
		panic("核心线程数应小于等于最大线程数")
	}
	// 2. 创建线程池
	tp := &ThreadPool{
		corePoolSize: corePoolSize,
		maxPoolSize:  maxPoolSize,
		taskQueue:    make(chan Task, queueSize),
		workerQueue:  make(chan struct{}, maxPoolSize),
		shutdown:     make(chan struct{}),
	}
	// 3. 创建核心线程
	for i := 0; i < corePoolSize; i ++ {
		// 3.1. 记录创建的线程，确保在关闭时不遗漏
		tp.wg.Add(1)
		// 3.2. 创建线程并启动
		go tp.worker()
	}

	return tp
}

func (tp *ThreadPool) worker() {
	// 1. 核心线程不停运行，并对任务队列和关闭信号进行监听
	defer tp.wg.Done()
	for {
		select {
		// 2. 若任务队列有任务，则取出任务并执行
		case task, ok := <-tp.taskQueue:
			if !ok {
				return
			}
			task()
		// 3. 若关闭信号发出，则退出线程
		case <-tp.shutdown:
			return
		}
	}
}

// submit 提交任务
func (tp *ThreadPool) submit(task Task) {
	select {
	// 1. 若线程池未关闭，则将任务放入任务队列
	case tp.taskQueue <- task:
	// 2. 若线程池已关闭，则丢弃任务
	case <-tp.shutdown:
		return
	// 3. 若任务队列已满，则创建新的线程
	default:
		select {
		case tp.workerQueue <- struct{}{}:
			tp.wg.Add(1)
			go tp.worker()
			tp.taskQueue <- task
		// 4. 无法创建新线程，任务被拒绝
		default:
			fmt.Println("Task rejected: queue is full and no more threads can be created")
		}
	}
}

// Shutdown 关闭线程池
func (p *ThreadPool) Shutdown() {
	close(p.shutdown)
	p.wg.Wait()
	close(p.taskQueue)
}

File: thread-pool.cpp
#include <atomic>
#include <condition_variable>
#include <functional>
#include <iostream>
#include <mutex>
#include <queue>
#include <thread>
#include <vector>

class ThreadPool {
 public:
    ThreadPool(size_t corePoolSize, size_t maxPoolSize, size_t queueSize)
        // 1. 初始化线程池基本参数
        : corePoolSize_(corePoolSize), maxPoolSize_(maxPoolSize), queueSize_(queueSize), isShutdown_(false) {
            // 2. 创建核心线程
            for (size_t i = 0; i < corePoolSize_; i ++) {
                workers_.emplace_back([this]() -> void { this->workerThread(); });
            }
        }

    ~ThreadPool() {
        Shutdown();
    }

    void Submit(std::function<void()> task) {
        std::unique_lock<std::mutex> lock(mtx_);
        if (isShutdown_) {
            std::cerr << "线程池已关闭，不能提交任务" << std::endl;
        }
        if (tasks_.size() >= queueSize_ && workers_.size() >= maxPoolSize_) {
            std::cerr << "已达到最大线程数，不能提交任务" << std::endl;
        }
        if (tasks_.size() >= queueSize_ && workers_.size() < maxPoolSize_) {
            workers_.emplace_back([this] { this->workerThread(); });
        }
        tasks_.push(task);
        lock.unlock();
        cv_.notify_one();
    }

    void Shutdown() {
        if (isShutdown_) {
            return;
        }
        isShutdown_ = true;
        cv_.notify_all();
        for (auto& worker : workers_) {
            if (worker.joinable()) {
                worker.join();
            }
        }
    }

 private:
    void workerThread() {
        while (!isShutdown_) {
            std::function<void()> task;
            std::unique_lock<std::mutex> lock(mtx_);
            cv_.wait(lock, [this]() -> bool { return !tasks_.empty() || isShutdown_; });
            if (isShutdown_ || tasks_.empty()) {
                return;
            }
            task = tasks_.front();
            tasks_.pop();
            lock.unlock();
            task();
        }
    }

    size_t corePoolSize_;                       // 核心线程数
    size_t maxPoolSize_;                        // 最大线程数
    size_t queueSize_;                          // 任务队列大小 
    std::vector<std::thread> workers_;          // 工作线程集合
    std::queue<std::function<void()>> tasks_;   // 任务队列
    std::mutex mtx_;                            // 互斥锁
    std::condition_variable cv_;                // 条件变量
    std::atomic<bool> isShutdown_;              // 线程池是否关闭
};
